cpp/src/arrow/util/decimal.h

```diff
@@ -0,0 +1,190 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef ARROW_DECIMAL_H
+#define ARROW_DECIMAL_H
+
+#include <cmath>
+#include <cstdlib>
+#include <iterator>
+#include <regex>
+#include <string>
+
+#include "arrow/status.h"
+#include "arrow/util/bit-util.h"
+#include "arrow/util/logging.h"
+
+#ifdef _MSC_VER
+#include <boost/multiprecision/cpp_int.hpp>
+#endif
+
+namespace arrow {
+
+#ifdef _MSC_VER
+using boost::multiprecision::int128_t;
+#else
+typedef __int128_t int128_t;
+#endif
+
+template <typename T>
+struct ARROW_EXPORT Decimal;
+
+static inline Status StringToInteger(
+    const std::string& whole, const std::string& fractional, int8_t sign, int128_t* out) {
+  // TODO(phillipc): There's probably a more intelligent way to do this
+  size_t i = 0;
+  *out = static_cast<int128_t>(0);
+  for (auto iter = fractional.crbegin(); iter != fractional.crend(); ++iter) {
```

Comment:
I'm going to refactor this to use boost multiprecision, which has conversion functions so there's no need for this.